Sporadic dependency errors when building in parallel (mvn -T ...)

Preface: Finding this issue was quite a journey because I first considered it to be a Maven core bug (https://github.com/apache/maven/pull/310), but the problem is actually caused by this plugin.

When building in parallel, compilation in a submodule of my current closed source project (sorry, cannot share!) fails _sporadically_ with errors like:
```
17:12:46  [ERROR] <someJavaFile>:[10,37] package org.jboss.resteasy.spi does not exist
17:12:46  [ERROR] <someJavaFile>:[10,1] static import only from classes and interfaces
17:12:46  [ERROR] <someJavaFile>:[11,37] package org.jboss.resteasy.spi does not exist
17:12:46  [ERROR] <someJavaFile>:[11,1] static import only from classes and interfaces
17:12:46  [ERROR] <someJavaFile>:[12,37] package org.jboss.resteasy.spi does not exist
17:12:46  [ERROR] <someJavaFile>:[12,1] static import only from classes and interfaces
17:12:46  [ERROR] <someJavaFile>:[14,24] package javax.ws.rs.core does not exist
```
The "missing" packages come from transitive dependencies of [`resteasy-core`](https://mvnrepository.com/artifact/org.jboss.resteasy/resteasy-core/4.5.3.Final) which in turn is a transitive dependency of [`quarkus-resteasy`](https://mvnrepository.com/artifact/io.quarkus/quarkus-resteasy).

It turns out that when the errors strike, there is a Warning further up in the log that is very easy to miss:
```
[WARNING] The POM for org.jboss.resteasy:resteasy-core:jar:4.5.3.Final is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
```
I modified the logging in `maven-core` (as `-X` would change the timing in a way I could not reproduce the problem anymore) so that I could see what was going on:
```
org.apache.maven.model.building.ModelBuildingException: 17 problems were encountered while building the effective model for org.jboss.resteasy:resteasy-core:4.5.3.Final
[ERROR] 'dependencies.dependency.version' for org.jboss.spec.javax.ws.rs:jboss-jaxrs-api_2.1_spec:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.spec.javax.xml.bind:jboss-jaxb-api_2.3_spec:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.spec.javax.servlet:jboss-servlet-api_4.0_spec:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.resteasy:resteasy-tracing-api:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.reactivestreams:reactive-streams:jar is missing. @
[ERROR] 'dependencies.dependency.version' for jakarta.validation:jakarta.validation-api:jar is missing. @
[ERROR] 'dependencies.dependency.version' for junit:junit:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.apache.logging.log4j:log4j-api:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.apache.logging.log4j:log4j-core:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.spec.javax.annotation:jboss-annotations-api_1.3_spec:jar is missing. @
[ERROR] 'dependencies.dependency.version' for com.sun.activation:jakarta.activation:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.spec.javax.el:jboss-el-api_3.0_spec:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.logging:jboss-logging:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.logging:jboss-logging-annotations:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.jboss.logging:jboss-logging-processor:jar is missing. @
[ERROR] 'dependencies.dependency.version' for org.eclipse.microprofile.config:microprofile-config-api:jar is missing. @
[ERROR] 'dependencies.dependency.version' for io.smallrye.config:smallrye-config:jar is missing. @

    at org.apache.maven.model.building.DefaultModelProblemCollector.newModelBuildingException (DefaultModelProblemCollector.java:197)
    at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:509)
```

Luckily, this reminded me of the following problem I ran into with an earlier version of this plugin: https://github.com/mojohaus/flatten-maven-plugin/issues/87#issuecomment-488781161

And indeed, there is a special dependency-defining profile involved in the parent of `resteasy-core`: https://repo1.maven.org/maven2/org/jboss/resteasy/resteasy-jaxrs-all/4.5.3.Final/resteasy-jaxrs-all-4.5.3.Final.pom
```xml
<profile>
    <id>resteasy-default</id>
    <activation>
        <property>
            <name>!resteasy.dependencies.eap</name>
        </property>
    </activation>
    <modules>
        <!-- ... -->
    </modules>
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.jboss.resteasy</groupId>
                <artifactId>resteasy-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
</profile>
```

So my theory is, that the dreaded profile handling of this plugin in the combination with the not-so-immutable and thus not-so-thread-safe Maven internals sometimes yields a state, where this profile is not activated and so the versions are missing.
Due to the complexity and volatile nature of the problem, I cannot actually pinpoint the exact location of the root cause.

At the moment I am testing a fix/workaround for this locally and so far it has been looking good!

PS: So it turns out that [my fears](https://github.com/mojohaus/flatten-maven-plugin/issues/87#issuecomment-536941275) in context of #87 were unfortunately right.

--- a/src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java
+++ b/src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java
@@ -33,9 +33,7 @@ import org.apache.maven.model.Plugin;
 import org.apache.maven.model.Profile;
 import org.apache.maven.model.Repository;
 import org.apache.maven.model.RepositoryPolicy;
-import org.apache.maven.model.building.DefaultModelBuilder;
 import org.apache.maven.model.building.DefaultModelBuildingRequest;
-import org.apache.maven.model.building.ModelBuilder;
 import org.apache.maven.model.building.ModelBuildingException;
 import org.apache.maven.model.building.ModelBuildingRequest;
 import org.apache.maven.model.building.ModelBuildingResult;
@@ -60,7 +58,6 @@ import org.apache.maven.shared.dependency.tree.DependencyTreeBuilderException;
 import org.apache.maven.shared.dependency.tree.traversal.DependencyNodeVisitor;
 import org.codehaus.mojo.flatten.cifriendly.CiInterpolator;
 import org.codehaus.mojo.flatten.model.resolution.FlattenModelResolver;
-import org.codehaus.plexus.util.IOUtil;
 import org.codehaus.plexus.util.StringUtils;
 import org.codehaus.plexus.util.xml.Xpp3Dom;
 import org.eclipse.aether.artifact.DefaultArtifact;
@@ -72,6 +69,7 @@ import org.xml.sax.Attributes;
 import org.xml.sax.SAXException;
 import org.xml.sax.ext.DefaultHandler2;
 
+import javax.inject.Inject;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 import java.io.File;
@@ -359,20 +357,14 @@ public class FlattenMojo
     @Component
     private DependencyResolver dependencyResolver;
 
-    @Component(role = ModelBuilder.class)
-    private DefaultModelBuilder defaultModelBuilder;
-    
     @Component( hint = "default" )
     private DependencyTreeBuilder dependencyTreeBuilder;
 
     @Component(role = ArtifactDescriptorReader.class)
     private ArtifactDescriptorReader artifactDescriptorReader;
 
-    @Component
-    private ProfileSelector profileSelector;
-
-    @Component
-    private ProfileInjector profileInjector;
+    @Inject
+    private ModelBuilderThreadSafetyWorkaround modelBuilderThreadSafetyWorkaround;
 
     /**
      * The constructor.
@@ -870,7 +862,7 @@ public class FlattenMojo
         ModelBuildingResult buildingResult;
         try
         {
-            ProfileInjector injector = new ProfileInjector()
+            ProfileInjector customInjector = new ProfileInjector()
             {
                 public void injectProfile( Model model, Profile profile, ModelBuildingRequest request,
                                            ModelProblemCollector problems )
@@ -885,7 +877,7 @@ public class FlattenMojo
                     }
                 }
             };
-            ProfileSelector selector = new ProfileSelector()
+            ProfileSelector customSelector = new ProfileSelector()
             {
                 public List<Profile> getActiveProfiles( Collection<Profile> profiles, ProfileActivationContext context,
                                                         ModelProblemCollector problems )
@@ -904,24 +896,8 @@ public class FlattenMojo
                     return activeProfiles;
                 }
             };
-            
-            // brief modification of singleton defaultModelBuilder needs to be limited to one thread/execution at a time
-            synchronized( defaultModelBuilder )
-            {
-                try
-                {
-                    defaultModelBuilder.setProfileInjector( injector ).setProfileSelector( selector );
-                    //if (flattenMode == FlattenMode.resolveCiFriendliesOnly) {
-                    //  defaultModelBuilder.setModelInterpolator(new CiModelInterpolator());
-                    //}
-                    buildingResult = defaultModelBuilder.build( buildingRequest );
-                }
-                finally
-                {
-                    // reset profileInjector and profileSelector
-                    defaultModelBuilder.setProfileInjector( this.profileInjector ).setProfileSelector( this.profileSelector );
-                }
-            }
+
+            buildingResult =  modelBuilderThreadSafetyWorkaround.build( buildingRequest, customInjector, customSelector );
         }
         catch ( ModelBuildingException e )
         {

--- a/src/main/java/org/codehaus/mojo/flatten/ModelBuilderThreadSafetyWorkaround.java
+++ b/src/main/java/org/codehaus/mojo/flatten/ModelBuilderThreadSafetyWorkaround.java
@@ -1,0 +1,132 @@
+package org.codehaus.mojo.flatten;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.apache.maven.model.building.DefaultModelBuilderFactory;
+import org.apache.maven.model.building.ModelBuildingException;
+import org.apache.maven.model.building.ModelBuildingRequest;
+import org.apache.maven.model.building.ModelBuildingResult;
+import org.apache.maven.model.building.ModelProcessor;
+import org.apache.maven.model.composition.DependencyManagementImporter;
+import org.apache.maven.model.inheritance.InheritanceAssembler;
+import org.apache.maven.model.interpolation.ModelInterpolator;
+import org.apache.maven.model.management.DependencyManagementInjector;
+import org.apache.maven.model.management.PluginManagementInjector;
+import org.apache.maven.model.normalization.ModelNormalizer;
+import org.apache.maven.model.path.ModelPathTranslator;
+import org.apache.maven.model.path.ModelUrlNormalizer;
+import org.apache.maven.model.plugin.LifecycleBindingsInjector;
+import org.apache.maven.model.plugin.PluginConfigurationExpander;
+import org.apache.maven.model.plugin.ReportConfigurationExpander;
+import org.apache.maven.model.plugin.ReportingConverter;
+import org.apache.maven.model.profile.ProfileInjector;
+import org.apache.maven.model.profile.ProfileSelector;
+import org.apache.maven.model.superpom.SuperPomProvider;
+import org.apache.maven.model.validation.ModelValidator;
+import org.eclipse.sisu.Nullable;
+
+/**
+ * Works around thread safety issues when modifying the global singleton {@link org.apache.maven.model.building.DefaultModelBuilder DefaultModelBuilder}
+ * with custom {@link ProfileInjector} and {@link ProfileSelector}. Instead of modifying the global {@code DefaultModelBuilder}, this class
+ * creates a new {@code DefaultModelBuilder} and equips it with the currently active components like {@link ModelProcessor}, {@link ModelValidator} etc.
+ * which might have been modified/provided by other Maven extensions. 
+ * 
+ * @author Falko Modler
+ * @since 1.2.3
+ */
+@Named
+public class ModelBuilderThreadSafetyWorkaround
+{
+
+    @Inject
+    private ModelProcessor modelProcessor;
+
+    @Inject
+    private ModelValidator modelValidator;
+
+    @Inject
+    private ModelNormalizer modelNormalizer;
+
+    @Inject
+    private ModelInterpolator modelInterpolator;
+
+    @Inject
+    private ModelPathTranslator modelPathTranslator;
+
+    @Inject
+    private ModelUrlNormalizer modelUrlNormalizer;
+
+    @Inject
+    private SuperPomProvider superPomProvider;
+
+    @Inject
+    private InheritanceAssembler inheritanceAssembler;
+
+    @Inject
+    private PluginManagementInjector pluginManagementInjector;
+
+    @Inject
+    private DependencyManagementInjector dependencyManagementInjector;
+
+    @Inject
+    private DependencyManagementImporter dependencyManagementImporter;
+
+    @Inject
+    @Nullable
+    private LifecycleBindingsInjector lifecycleBindingsInjector;
+
+    @Inject
+    private PluginConfigurationExpander pluginConfigurationExpander;
+
+    @Inject
+    private ReportConfigurationExpander reportConfigurationExpander;
+
+    @Inject
+    private ReportingConverter reportingConverter;
+
+    public ModelBuildingResult build( ModelBuildingRequest buildingRequest, ProfileInjector customInjector, ProfileSelector customSelector )
+        throws ModelBuildingException
+    {
+        // note: there is neither DefaultModelBuilder.get*(), nor DefaultModelBuilder.clone()
+        return new DefaultModelBuilderFactory().newInstance()
+            .setProfileInjector( customInjector )
+            .setProfileSelector( customSelector )
+            // apply currently active ModelProcessor etc. to support extensions like jgitver
+            .setDependencyManagementImporter( dependencyManagementImporter )
+            .setDependencyManagementInjector( dependencyManagementInjector )
+            .setInheritanceAssembler( inheritanceAssembler )
+            .setLifecycleBindingsInjector( lifecycleBindingsInjector )
+            .setModelInterpolator( modelInterpolator )
+            .setModelNormalizer( modelNormalizer )
+            .setModelPathTranslator( modelPathTranslator )
+            .setModelProcessor( modelProcessor )
+            .setModelUrlNormalizer( modelUrlNormalizer )
+            .setModelValidator( modelValidator )
+            .setPluginConfigurationExpander( pluginConfigurationExpander )
+            .setPluginManagementInjector( pluginManagementInjector )
+            .setReportConfigurationExpander( reportConfigurationExpander )
+            .setReportingConverter( reportingConverter )
+            .setSuperPomProvider( superPomProvider )
+            .build( buildingRequest );
+    }
+}

--- a/src/test/java/org/codehaus/mojo/flatten/CreateEffectivePomTest.java
+++ b/src/test/java/org/codehaus/mojo/flatten/CreateEffectivePomTest.java
@@ -30,7 +30,11 @@ import org.apache.maven.artifact.repository.layout.DefaultRepositoryLayout;
 import org.apache.maven.model.Model;
 import org.apache.maven.model.building.DefaultModelBuilderFactory;
 import org.apache.maven.model.building.DefaultModelBuildingRequest;
+import org.apache.maven.model.building.ModelBuildingException;
 import org.apache.maven.model.building.ModelBuildingRequest;
+import org.apache.maven.model.building.ModelBuildingResult;
+import org.apache.maven.model.profile.ProfileInjector;
+import org.apache.maven.model.profile.ProfileSelector;
 import org.apache.maven.project.DefaultProjectBuildingRequest;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.shared.dependencies.resolve.internal.DefaultDependencyResolver;
@@ -84,7 +88,7 @@ public class CreateEffectivePomTest
                 depencencyResolver, projectBuildingRequest, Collections.<MavenProject>emptyList() );
         ModelBuildingRequest buildingRequest =
             new DefaultModelBuildingRequest().setPomFile( pomFile ).setModelResolver( resolver ).setUserProperties( userProperties );
-        setDeclaredField(tested, "defaultModelBuilder", new DefaultModelBuilderFactory().newInstance());
+        setDeclaredField( tested, "modelBuilderThreadSafetyWorkaround", buildModelBuilderThreadSafetyWorkaroundForTest() );
         Model effectivePom = tested.createEffectivePom( buildingRequest, false, FlattenMode.defaults );
         assertThat( effectivePom.getName() ).isEqualTo( magicValue );
     }
@@ -95,5 +99,23 @@ public class CreateEffectivePomTest
         Field field = pojo.getClass().getDeclaredField( fieldName );
         field.setAccessible( true );
         field.set( pojo, propertyValue );
+    }
+
+    /**
+     * @return ModelBuilderThreadSafetyWorkaround with a reduced scope for this simple test
+     */
+    private ModelBuilderThreadSafetyWorkaround buildModelBuilderThreadSafetyWorkaroundForTest()
+    {
+        return new ModelBuilderThreadSafetyWorkaround() {
+            @Override
+            public ModelBuildingResult build(ModelBuildingRequest buildingRequest, ProfileInjector customInjector, ProfileSelector customSelector)
+                    throws ModelBuildingException {
+
+                return new DefaultModelBuilderFactory().newInstance()
+                    .setProfileInjector( customInjector )
+                    .setProfileSelector( customSelector )
+                    .build( buildingRequest );
+            }
+        };
     }
 }

